src/sa.py

```diff
@@ -1,15 +1,98 @@
 import argparse
+import fasta, fastq
+import re
 
+def getSuffixes(x: memoryview):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    return suffixes
 
+def getTrailingNumber(s):
+    m = re.search(r'\d+$', s)
+    return (s[:m.start()], int(s[m.start():]))
+
+def lower(a: str, x: str, sa: memoryview, lo: int, hi: int, offset: int) -> int:
+    """Finds the lower bound of `a` at `offset` in the block defined by `lo:hi`."""
+    while lo < hi:  # Search in sa[lo:hi]
+        m = (lo + hi) // 2
+        if x[sa[m] + offset % len(x)] < a: # compare at column offset in sa
+            lo = m + 1
+        else:
+            hi = m
+    return lo
+
+def upper(a: str, x: str, sa: memoryview, lo: int, hi: int, offset: int) -> int:
+    """Finds the upper bound of `a` at `offset` in the block defined by `lo:hi`."""
+    return lower(chr(ord(a) + 1), x, sa, lo, hi, offset)
+
+def search(sa, pattern, genome):
+    lo, hi = 0, len(sa)
+    for offset, a in enumerate(pattern):
+        lo = lower(a, genome, sa, lo, hi, offset)
+        hi = upper(a, genome, sa, lo, hi, offset)
+    for sol in sa[lo:hi]:
+        yield sol+1
+    return
```

Comment:
You should also be able to `yield from sa[lo:hi]` instead of the loop. The `return` doesn't do anything, so you might as well save six bytes by deleting it.